### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by mahshidmansouri**

### CODE REVIEW

**Comment by mahshidmansouri** [comment]
This code is for pull request.

**Comment by golpashin** [comment]
Hi Mahshid, I'll review your code for this HW! :D

**Comment by golpashin** [comment]
Hi Mahshid, here are my comments. I'll try to add a few more later if I come across something.

Overall, I think the code is well-written; the provided comments are helpful in figuring out what every function or command is doing. I also appreciate that you have provided and explained your hyperparameters at the top of the code.

Here are my suggestions regarding the rest of the code:
-Instead of target_net.state_dict = q_net.state_dict you should use target_net.load_state_dict(q_net.state_dict()).
-The c variable used to update the target network can be defined within the loop where it is used, rather than outside it.
-I think you are saving the Q-Network parameters after each episode. It’d be more efficient to save the parameters only after the training is complete, i.e. after the last episode.
-For your helpers.py code, I suggest you use “optimize_Q” instead of the “optimize_model” name.
-I suggest adding individual comments for optimizer.zero_grad(), loss.backward(), and, ptimizer.step() explaining what each one does.
-In plot_state_value_function function, you can replace the nested for-loops with meshgrid function to create a grid of x-y coordinates.

**Comment by mahshidmansouri** [comment]
Hi Alen, 

Thanks for your comments. I've addressed all your comments. The only comment that I have a followup question about is your third comment saying: "-I think you are saving the Q-Network parameters after each episode. It’d be more efficient to save the parameters only after the training is complete, i.e. after the last episode.". I guess based on my current code, I'm saving the trained Q-Network parameters at the end of each episode which is after the for-loop. Please correct me if I am wrong. 

Thank you! 
Mahshid

**Comment by golpashin** [comment]
HI Mahshid, 
I think what I was referring to was that you are calling 'c += 1'  and the following code within the 'while not done' loop:
if (with_TARGET_Q): 
                if (c % target_update_freq == 0):
                    target_net.state_dict = q_net.state_dict

This condition is checked at every step of every trajectory 'i_epsiode'.  So basically, at every c=20 steps within each i_epsiode, the weights are updated. My suggestion was to move the 'if (with_TARGET_Q): ' out of the 'while not done' loop, under the 'for i_episode in range(num_episodes):'. Doing so, will only reset the parameters after every 20 i_epsiode's. This would make the NN weight update much slower. Let me know what you think.

______________
Now scrolling through the code, I cam across one more thing I wanted to ask. When you initialize the optimizer, you have lr=0.00025. I might be wrong here since I haven't used Python for optimization before, but isn't 'lr' the learning rate? If so, shouldn't lr=alpha? The numerical 'lr' value looks right, but I am not sure what 'alpha' and the other 'momentum' options are.   

optimizer = optim.RMSprop(q_net.parameters(),lr=0.00025, alpha=0.99, eps=1e-08, weight_decay=0, momentum=0.95)


thanks!

Alen

**Comment by golpashin** [comment]
For the first comment, I meant to suggest moving the  'if (with_TARGET_Q): '  as well as the 'c += 1'  commands out of the 'while not done' loop. Sorry for the confusion.

**Comment by mahshidmansouri** [comment]
Hi Alen, 

Thanks for these suggestions! They were really helpful. I have reflected these suggestions under my updated code. 

You're totally right about my target network update rate. I moved that block of code to outside the while-loop to ensure that I'm updating my target network every certain number of episodes. 

I also changed my optimizer parameters to what you suggested. 

Thank you!
Mahshid
